Use the ArrayPool API in .NET Standard 2.1 or later to reduce the creation of LOH objects.

Wak1Z2 opened the pull request:
Hi, I'm a programmer from China. The text below is translated using AI, as my English isn't proficient enough for direct conversation, though I'm actively working to improve.

I'm developing a personal web application using Magick.NET. During development, I noticed that converting RAW photo files to JPG generates a significant number of LOH (Large Object Heap) objects. After reviewing the source code, I traced the issue to the ResizeBytes method in the ByteArrayWrapper class. This method reallocates byte arrays based on the length returned by localized code, causing unused arrays to accumulate in the LOH. To address this, I implemented the ArrayPool API to reduce LOH allocations. After the modifications, the program now generates only a small number of LOH objects per run, drastically reducing LOH allocation.

My environment uses .NET 8, and the ArrayPool API is only available in .NET Standard 2.1 or later, meaning it cannot compile in versions prior to .NET Standard 2.0 (e.g., .NET Framework 4.6.1).

Below are screenshots of how my interface uses Magick.NET and the memory monitoring. Each LOH change in the screenshots corresponds to one program execution.

![eca6fe889feeb0061d073ceab97fcf34](https://github.com/user-attachments/assets/cc2e3a39-c1e5-4365-9625-4d14469ff2b5)

Before：
![before1](https://github.com/user-attachments/assets/d7f05ffe-f6bb-42e0-815b-efdc0b5d992b)
![before2](https://github.com/user-attachments/assets/effd6ef6-b27c-486a-a595-5c4ea942ede1)
After：
![after1](https://github.com/user-attachments/assets/442375ee-b2e2-42a2-b7b1-63466ac99749)
![after2](https://github.com/user-attachments/assets/c9bce2cf-4ebb-4722-ae0d-bfec833d568f)
![after3](https://github.com/user-attachments/assets/52ddba2f-3d43-4211-86cb-4bc9059b35b4)

Here are my test photos, captured using a Nikon Z8 in uncompressed RAW format.
https://github.com/Wak1Z2/Magick.NET/tree/main/testImage

Thank you for reviewing my code! This is my first attempt at contributing to an open-source library.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/dlemstra/Magick.NET?pullRequest=1812) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/dlemstra/Magick.NET?pullRequest=1812) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/dlemstra/Magick.NET?pullRequest=1812) it.</sub>

dlemstra commented:
Thanks for your pull request and detailed explanation. I am right now traveling so it might take a while before I have time to review this.

Wak1Z2 commented:
> Thanks for your pull request and detailed explanation. I am right now traveling so it might take a while before I have time to review this.

Okay, wish you a pleasant journey~

dlemstra commented:
I have allowed the build a while back and it looks like your changes failed to build. Can you take a look at that?

dlemstra commented:
Feel free to reach out if you need more help with this PR.

Wak1Z2 commented:
My apologies — I’ve been too caught up in gaming recently and haven’t checked GitHub or emails. I’ll review them as soon as possible. Thank you for your patience!😂😂

dlemstra commented:
Take all the time that you need, just wanted to make sure you have read my message.

Wak1Z2 commented:
Thanks for the reminder! I've deleted the test photos and submitted the updated code. It’s 4 AM here — calling it a day. Wishing you a wonderful day ahead! 🌞

Wak1Z2 commented:
Good afternoon! I’ve submitted the code modifications and removed the test images as discussed. Is there anything else I need to address? This is my first time contributing to a public repository, so please let me know if I’ve missed any steps or requirements. 😊

Wak1Z2 commented:
I’ve reviewed the feedback and will work on addressing the reported issues in the coming days. Thank you for the thorough review! 😄

Wak1Z2 commented:
After reviewing the building.md documentation, I re-ran localized tests on my code. The issue lies in the TheFeaturesProperty.ContainsExpectedFeatures method: when I locally retrieve MagickNET.Features, the value does not include 'Debug'. After removing 'Debug' from the expected values, all tests passed. However, I couldn’t reproduce the GitHub CI failure locally. Could you please investigate this? I’ve uploaded the test log file for reference.
[output.log](https://github.com/user-attachments/files/19765227/output.log)

dlemstra commented:
I suspect that your branch is not up to to date with the `main` branch. You will to do a rebase or a merge.

Wak1Z2 commented:
Oh! Didn’t realize that — I’ve already synced my branch with the latest main branch. Thanks for the heads-up! 😄

dlemstra commented on src/Magick.NET/Helpers/ByteArrayWrapper.cs:
```diff
@@ -45,42 +87,53 @@ public long Read(IntPtr data, UIntPtr count, IntPtr user_data)
 
     public long Tell(IntPtr user_data)
-        => _offset;
+    {
```
Can you please revert these kind of code style changes? You are now making changes so that the coding style is not the same as the rest of this project.

Wak1Z2 replied:
The code style has been updated to align with your approach.

Wak1Z2 replied:
If any parts haven’t been updated yet, please highlight them. Thanks! 😊

dlemstra commented on src/Magick.NET/Helpers/ByteArrayWrapper.cs:
```diff
@@ -45,42 +87,53 @@ public long Read(IntPtr data, UIntPtr count, IntPtr user_data)
     public long Write(IntPtr data, UIntPtr count, IntPtr user_data)
     {
         if (data == IntPtr.Zero)
+        {
```
That also means these kind of changes. Please revert them.

Wak1Z2 replied:
Reverted.😁

dlemstra commented on src/Magick.NET/Helpers/ByteArrayWrapper.cs:
```diff
@@ -45,42 +87,53 @@ public long Read(IntPtr data, UIntPtr count, IntPtr user_data)
         EnsureLength(newOffset);
 
         var source = (byte*)data.ToPointer();
+
```
And these kind of changes that make it more difficult to see what was actually changed.

Wak1Z2 replied:
Reverted.😁

dlemstra commented on src/Magick.NET/Helpers/ByteArrayWrapper.cs:
```diff
@@ -94,17 +147,37 @@ public long Write(IntPtr data, UIntPtr count, IntPtr user_data)
+
+    private void ResizeBytes(int length)
+    {
+        byte[] byte2 = _pool.Rent(length);
```
This should have raised a warning because this project use `var` for the variable declaration instead of `byte[]`. And can you rename `byte2` to `newBytes` to make it more clear what is happening. But this part of the code is not compiled because the check for `NETSTANDARD2_1_OR_GREATER` doesn't work.

Wak1Z2 replied:
I’ve made the changes. Please review.

dlemstra commented on src/Magick.NET/Helpers/ByteArrayWrapper.cs:
```diff
@@ -3,34 +3,76 @@
 
 using System;
 using System.IO;
+#if NETSTANDARD2_1_OR_GREATER
```
This check should be `#if !NETSTANDARD2_0` instead to be in line with the rest of the project.

dlemstra commented on src/Magick.NET/Helpers/ByteArrayWrapper.cs:
```diff
@@ -45,42 +87,53 @@ public long Read(IntPtr data, UIntPtr count, IntPtr user_data)
-            SeekOrigin.End => _length - offset,
-            _ => -1,
-        });
+                                  {
```
Please revert this whitespace change.

Wak1Z2 replied:
Reverted.😁

dlemstra commented on src/Magick.NET/Helpers/ByteArrayWrapper.cs:
```diff
@@ -3,34 +3,76 @@
 {
+    #if NETSTANDARD2_1_OR_GREATER
+
+    private static readonly ArrayPool<byte> _pool = ArrayPool<byte>.Create(1024 * 1024 * 64, 1024);
```
I wonder if we should move this logic to a separate class so we can also use this in other places but I will do that at a later moment myself.

dlemstra commented on src/Magick.NET/Helpers/ByteArrayWrapper.cs:
```diff
@@ -3,34 +3,76 @@
+    
+~ByteArrayWrapper()
+    {
+        if (_bytes is not null)
```
How is it possible that `_bytes` is `null`?

dlemstra commented on src/Magick.NET/Helpers/ByteArrayWrapper.cs:
```diff
@@ -3,34 +3,76 @@
+
+    private static readonly ArrayPool<byte> _pool = ArrayPool<byte>.Create(1024 * 1024 * 64, 1024);
+
+    private byte[] _bytes = _pool.Rent(1024 * 1024);
```
Why did you increment the initial size of this buffer?

Wak1Z2 replied:
The _bytes cache size was arbitrarily set to 1024*1024 (1MB) due to my oversight. My sincere apologies! 🙏

dlemstra commented on src/Magick.NET/Helpers/ByteArrayWrapper.cs:
```diff
@@ -3,34 +3,76 @@
+
+    private int _length;
+
+    #if NETSTANDARD2_1_OR_GREATER
```
In the rest of the project this `#if` is at the start of the line. Can you please change this.

dlemstra commented on src/Magick.NET/Helpers/ByteArrayWrapper.cs:
```diff
@@ -94,17 +147,37 @@ public long Write(IntPtr data, UIntPtr count, IntPtr user_data)
+    private void ResizeBytes(int length)
+    {
+        byte[] byte2 = _pool.Rent(length);
+        Array.Copy(_bytes, byte2, Math.Min(_bytes.Length, length));
```
Why did you add the `Math.Min` here? And should this not always be `_length`?

Wak1Z2 replied:
I verified the relevant code and confirmed your feedback was correct. The fix has been included in the latest commit.

dlemstra commented on src/Magick.NET/Helpers/ByteArrayWrapper.cs:
```diff
@@ -94,17 +147,37 @@ public long Write(IntPtr data, UIntPtr count, IntPtr user_data)
+    {
+        byte[] byte2 = _pool.Rent(length);
+        Array.Copy(_bytes, byte2, Math.Min(_bytes.Length, length));
+        _pool.Return(_bytes, true);
```
Do we really need to clear the array here?

Wak1Z2 replied:
no,you are right😂

dlemstra commented on src/Magick.NET/Helpers/ByteArrayWrapper.cs:
```diff
@@ -3,34 +3,76 @@
+    public byte[] GetBytes()
+    {
+        var result = new byte[_length];
+        Array.Copy(_bytes, result, Math.Min(_bytes.Length, _length));
```
This should always be `_length`?

Wak1Z2 replied:
I’ve revised the method,please review the updated version.🥰

dlemstra reviewed (commented):
Thanks for making this change. I finally had some time to do a proper review. Feel free to ask for more information when my question or comment is not clear for you.

Wak1Z2 commented:
I have modified the code according to your comments and verified the issues you raised — they were indeed correct. Additionally, I noticed that the original GetBytes method can be called repeatedly on the same instance without losing its internal _bytes data, whereas my implementation causes the _bytes within the instance to potentially be overwritten by other instances. Therefore, I removed the _pool.Return method and delegated the _pool.Return logic to the finalizer. I’m unsure if this approach is correct. If you have a better solution, please advise. Thank you!

Wak1Z2 commented:
After a day of consideration, I decided to add it back. Currently, this class is not reused in the program. Therefore, it can be understood that GetBytes is called only once, and we'll address any future issues later.

dlemstra commented on src/Magick.NET/Helpers/ByteArrayWrapper.cs:
```diff
@@ -3,21 +3,45 @@
+    {
+        var result = new byte[_length];
+        Array.Copy(_bytes, result, _length);
+        _pool.Return(_bytes);
```
You should not return the `_bytes` to the pool here because you will then also do that again inside the destructor. So please remove this line and also remove the `true` argument in the destructor.

p.s. Thanks for your patience and I think we are getting close to getting this merged.

Wak1Z2 replied:
Thank you for your guidance. I have updated the code. If there are still any changes required or areas where I need to collaborate, please let me know.
